File: query-compiler-llvm/src/parser/lexer.cpp
#include "parser/lexer.h"
#include <algorithm>
#include <cctype>
#include <stdexcept>

namespace qc {

static std::string to_upper(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), ::toupper);
    return s;
}

static const std::unordered_map<std::string, TokenKind> KEYWORDS = {
    {"SELECT",   TokenKind::KW_SELECT},  {"FROM",     TokenKind::KW_FROM},
    {"WHERE",    TokenKind::KW_WHERE},   {"JOIN",     TokenKind::KW_JOIN},
    {"INNER",    TokenKind::KW_INNER},   {"LEFT",     TokenKind::KW_LEFT},
    {"RIGHT",    TokenKind::KW_RIGHT},   {"FULL",     TokenKind::KW_FULL},
    {"OUTER",    TokenKind::KW_OUTER},   {"ON",       TokenKind::KW_ON},
    {"AS",       TokenKind::KW_AS},      {"AND",      TokenKind::KW_AND},
    {"OR",       TokenKind::KW_OR},      {"NOT",      TokenKind::KW_NOT},
    {"BETWEEN",  TokenKind::KW_BETWEEN}, {"IN",       TokenKind::KW_IN},
    {"IS",       TokenKind::KW_IS},      {"NULL",     TokenKind::KW_NULL},
    {"TRUE",     TokenKind::KW_TRUE},    {"FALSE",    TokenKind::KW_FALSE},
    {"DISTINCT", TokenKind::KW_DISTINCT},{"GROUP",    TokenKind::KW_GROUP},
    {"BY",       TokenKind::KW_BY},      {"HAVING",   TokenKind::KW_HAVING},
    {"ORDER",    TokenKind::KW_ORDER},   {"ASC",      TokenKind::KW_ASC},
    {"DESC",     TokenKind::KW_DESC},    {"LIMIT",    TokenKind::KW_LIMIT},
    {"OFFSET",   TokenKind::KW_OFFSET},  {"NULLS",    TokenKind::KW_NULLS},
    {"FIRST",    TokenKind::KW_FIRST},   {"LAST",     TokenKind::KW_LAST},
    {"DATE",     TokenKind::KW_DATE},    {"CAST",     TokenKind::KW_CAST},
    {"COUNT",    TokenKind::KW_COUNT},   {"SUM",      TokenKind::KW_SUM},
    {"AVG",      TokenKind::KW_AVG},     {"MIN",      TokenKind::KW_MIN},
    {"MAX",      TokenKind::KW_MAX},
};

Lexer::Lexer(std::string input) : input_(std::move(input)) {}

void Lexer::skip_whitespace() {
    while (pos_ < input_.size()) {
        char c = input_[pos_];
        if (c == '\n') { line_++; pos_++; }
        else if (std::isspace(c)) { pos_++; }
        else if (c == '-' && pos_ + 1 < input_.size() && input_[pos_+1] == '-') {
            // Line comment
            while (pos_ < input_.size() && input_[pos_] != '\n') pos_++;
        }
        else break;
    }
}

Token Lexer::read_number() {
    size_t start = pos_;
    bool is_float = false;
    while (pos_ < input_.size() && std::isdigit(input_[pos_])) pos_++;
    if (pos_ < input_.size() && input_[pos_] == '.') {
        is_float = true;
        pos_++;
        while (pos_ < input_.size() && std::isdigit(input_[pos_])) pos_++;
    }
    return {is_float ? TokenKind::FLOAT_LIT : TokenKind::INT_LIT,
            input_.substr(start, pos_ - start), line_};
}

Token Lexer::read_string() {
    pos_++; // skip opening quote
    size_t start = pos_;
    while (pos_ < input_.size() && input_[pos_] != '\'') {
        if (input_[pos_] == '\n') line_++;
        pos_++;
    }
    std::string text = input_.substr(start, pos_ - start);
    if (pos_ < input_.size()) pos_++; // skip closing quote
    return {TokenKind::STRING_LIT, text, line_};
}

Token Lexer::read_ident_or_keyword() {
    size_t start = pos_;
    while (pos_ < input_.size() && (std::isalnum(input_[pos_]) || input_[pos_] == '_'))
        pos_++;
    std::string text = input_.substr(start, pos_ - start);
    std::string upper = to_upper(text);
    auto it = KEYWORDS.find(upper);
    if (it != KEYWORDS.end())
        return {it->second, upper, line_};
    return {TokenKind::IDENT, text, line_};
}

Token Lexer::read_token() {
    skip_whitespace();
    if (pos_ >= input_.size())
        return {TokenKind::END_OF_FILE, "", line_};

    char c = input_[pos_];

    if (std::isdigit(c)) return read_number();
    if (c == '\'')       return read_string();
    if (std::isalpha(c) || c == '_') return read_ident_or_keyword();

    pos_++;
    switch (c) {
    case '(': return {TokenKind::LPAREN,    "(", line_};
    case ')': return {TokenKind::RPAREN,    ")", line_};
    case ',': return {TokenKind::COMMA,     ",", line_};
    case '.': return {TokenKind::DOT,       ".", line_};
    case ';': return {TokenKind::SEMICOLON, ";", line_};
    case '*': return {TokenKind::STAR,      "*", line_};
    case '+': return {TokenKind::PLUS,      "+", line_};
    case '-': return {TokenKind::MINUS,     "-", line_};
    case '/': return {TokenKind::SLASH,     "/", line_};
    case '%': return {TokenKind::PERCENT,   "%", line_};
    case '=': return {TokenKind::EQ,        "=", line_};
    case '<':
        if (pos_ < input_.size() && input_[pos_] == '=') { pos_++; return {TokenKind::LE, "<=", line_}; }
        if (pos_ < input_.size() && input_[pos_] == '>') { pos_++; return {TokenKind::NEQ,"<>", line_}; }
        return {TokenKind::LT, "<", line_};
    case '>':
        if (pos_ < input_.size() && input_[pos_] == '=') { pos_++; return {TokenKind::GE, ">=", line_}; }
        return {TokenKind::GT, ">", line_};
    case '!':
        if (pos_ < input_.size() && input_[pos_] == '=') { pos_++; return {TokenKind::NEQ,"!=", line_}; }
        break;
    }
    throw std::runtime_error(std::string("Unexpected character: ") + c);
}

Token Lexer::next() {
    if (has_peek_) {
        has_peek_ = false;
        return peek_token_;
    }
    return read_token();
}

Token Lexer::peek() {
    if (!has_peek_) {
        peek_token_ = read_token();
        has_peek_ = true;
    }
    return peek_token_;
}

void Lexer::consume() { (void)next(); }

} // namespace qc

File: volcano-push-adaptive-engine/src/adaptive_engine/sql/lexer.py
"""SQL lexer — converts a SQL string into a flat token stream."""
from __future__ import annotations
import re
from dataclasses import dataclass
from enum import Enum, auto


class TT(Enum):
    """Token type."""
    # Literals
    INTEGER   = auto()
    FLOAT     = auto()
    STRING    = auto()
    TRUE      = auto()
    FALSE     = auto()
    NULL      = auto()
    # Identifiers / keywords
    IDENT     = auto()
    # Keywords (each gets its own type for unambiguous parsing)
    SELECT    = auto()
    FROM      = auto()
    WHERE     = auto()
    JOIN      = auto()
    INNER     = auto()
    LEFT      = auto()
    RIGHT     = auto()
    FULL      = auto()
    OUTER     = auto()
    ON        = auto()
    AND       = auto()
    OR        = auto()
    NOT       = auto()
    AS        = auto()
    GROUP     = auto()
    BY        = auto()
    ORDER     = auto()
    HAVING    = auto()
    LIMIT     = auto()
    OFFSET    = auto()
    ASC       = auto()
    DESC      = auto()
    IS        = auto()
    IN        = auto()
    BETWEEN   = auto()
    LIKE      = auto()
    DISTINCT  = auto()
    STAR      = auto()
    # Aggregate / scalar functions
    COUNT     = auto()
    SUM       = auto()
    AVG       = auto()
    MIN       = auto()
    MAX       = auto()
    # Operators
    EQ        = auto()   # =
    NEQ       = auto()   # != or <>
    LT        = auto()   # <
    LTE       = auto()   # <=
    GT        = auto()   # >
    GTE       = auto()   # >=
    PLUS      = auto()   # +
    MINUS     = auto()   # -
    SLASH     = auto()   # /
    # Punctuation
    COMMA     = auto()
    DOT       = auto()
    LPAREN    = auto()
    RPAREN    = auto()
    SEMICOLON = auto()
    # Control
    EOF       = auto()


_KEYWORDS: dict[str, TT] = {
    "select": TT.SELECT, "from": TT.FROM, "where": TT.WHERE,
    "join": TT.JOIN, "inner": TT.INNER, "left": TT.LEFT,
    "right": TT.RIGHT, "full": TT.FULL, "outer": TT.OUTER,
    "on": TT.ON, "and": TT.AND, "or": TT.OR, "not": TT.NOT,
    "as": TT.AS, "group": TT.GROUP, "by": TT.BY, "order": TT.ORDER,
    "having": TT.HAVING, "limit": TT.LIMIT, "offset": TT.OFFSET,
    "asc": TT.ASC, "desc": TT.DESC, "is": TT.IS, "in": TT.IN,
    "between": TT.BETWEEN, "like": TT.LIKE, "distinct": TT.DISTINCT,
    "true": TT.TRUE, "false": TT.FALSE, "null": TT.NULL,
    "count": TT.COUNT, "sum": TT.SUM, "avg": TT.AVG,
    "min": TT.MIN, "max": TT.MAX,
}

_AGG_FUNCS: set[TT] = {TT.COUNT, TT.SUM, TT.AVG, TT.MIN, TT.MAX}


@dataclass(slots=True)
class Token:
    type: TT
    value: object  # str | int | float | None
    pos: int       # character offset in source

    def __repr__(self) -> str:
        return f"Token({self.type.name}, {self.value!r})"


class LexError(Exception):
    pass


def tokenize(sql: str) -> list[Token]:
    tokens: list[Token] = []
    i = 0
    n = len(sql)

    while i < n:
        # Skip whitespace
        if sql[i].isspace():
            i += 1
            continue

        # Single-line comment
        if sql[i : i + 2] == "--":
            while i < n and sql[i] != "\n":
                i += 1
            continue

        # String literal  'hello'
        if sql[i] == "'":
            j = i + 1
            while j < n and sql[j] != "'":
                if sql[j] == "\\" and j + 1 < n:
                    j += 1
                j += 1
            if j >= n:
                raise LexError(f"Unterminated string at position {i}")
            tokens.append(Token(TT.STRING, sql[i + 1 : j], i))
            i = j + 1
            continue

        # Number
        if sql[i].isdigit() or (sql[i] == "-" and i + 1 < n and sql[i + 1].isdigit()
                                  and (not tokens or tokens[-1].type not in (TT.INTEGER, TT.FLOAT, TT.IDENT, TT.RPAREN))):
            j = i
            if sql[j] == "-":
                j += 1
            while j < n and sql[j].isdigit():
                j += 1
            is_float = j < n and sql[j] == "."
            if is_float:
                j += 1
                while j < n and sql[j].isdigit():
                    j += 1
            raw = sql[i:j]
            tokens.append(Token(TT.FLOAT if is_float else TT.INTEGER,
                                 float(raw) if is_float else int(raw), i))
            i = j
            continue

        # Identifier or keyword
        if sql[i].isalpha() or sql[i] == "_":
            j = i
            while j < n and (sql[j].isalnum() or sql[j] == "_"):
                j += 1
            word = sql[i:j]
            tt = _KEYWORDS.get(word.lower(), TT.IDENT)
            tokens.append(Token(tt, word, i))
            i = j
            continue

        # Two-char operators
        two = sql[i : i + 2]
        if two == "!=":
            tokens.append(Token(TT.NEQ, "!=", i)); i += 2; continue
        if two == "<>":
            tokens.append(Token(TT.NEQ, "<>", i)); i += 2; continue
        if two == "<=":
            tokens.append(Token(TT.LTE, "<=", i)); i += 2; continue
        if two == ">=":
            tokens.append(Token(TT.GTE, ">=", i)); i += 2; continue

        # Single-char
        ONE = {
            "=": TT.EQ, "<": TT.LT, ">": TT.GT,
            "+": TT.PLUS, "-": TT.MINUS, "*": TT.STAR, "/": TT.SLASH,
            ",": TT.COMMA, ".": TT.DOT,
            "(": TT.LPAREN, ")": TT.RPAREN, ";": TT.SEMICOLON,
        }
        if sql[i] in ONE:
            tokens.append(Token(ONE[sql[i]], sql[i], i)); i += 1; continue

        raise LexError(f"Unexpected character {sql[i]!r} at position {i}")

    tokens.append(Token(TT.EOF, None, n))
    return tokens


# Re-export for convenience
AGG_FUNCS = _AGG_FUNCS
